fd/fd.c

```diff
@@ -680,62 +674,82 @@ M0_INTERNAL int m0_fd__tree_level_populate(const struct m0_conf_pver *pv,
 	return	M0_ERR(rc);
 }
 
-M0_INTERNAL int m0_fd__perm_cache_build(struct m0_fd_tree *tree)
+M0_INTERNAL int m0_fd_cache_grid_build(struct m0_layout *layout,
+				       struct m0_pdclust_instance *pi)
 {
-	struct m0_fd__tree_cursor  cursor;
-	struct m0_fd_tree_node    *node;
-	uint64_t                   children_nr;
-	uint64_t                  *cache_len;
-	uint64_t                  *cache_info;
-	uint64_t                   level;
-	int                        rc;
-
-	M0_ALLOC_ARR(cache_info, tree->ft_cnt);
-	if (cache_info == NULL)
+	struct m0_pool_version        *pver;
+	struct m0_fd_tree             *tree;
+	struct m0_fd_perm_cache_grid  *cache_grid;
+	struct m0_fd_perm_cache       *cache;
+	struct m0_fd__tree_cursor      cursor;
+	struct m0_fd_tree_node        *node;
+	uint32_t                       nxt_lvl_nodes;
+	uint32_t                       level;
+	int                            rc;
+
+	M0_PRE(layout != NULL && pi != NULL && layout->l_pver != NULL);
+
+	pver = layout->l_pver;
+	tree = &pver->pv_fd_tree;
+	cache_grid = m0_alloc(sizeof cache_grid[0]);
```

Comment:
cache_grid = m0_alloc(sizeof *cache_grid);

"cache_grid[0]" seems confusing. It "looks" like an array.

Reply:
Agree, fixed.